File: Qin/Qin-Android/app/src/main/java/edu/bfa/ss/qin/Util/ReBuildingRoom.java
package edu.bfa.ss.qin.Util;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import io.realm.Realm;

import static edu.bfa.ss.qin.Util.DatabaseHelper.LeancloudAPIBaseURL;

public class ReBuildingRoom {
    public static void CreateBuildingRoom(boolean refresh) {
        Building.GetAll(true);
        Room.GetAll(true);
        Realm realm = Realm.getDefaultInstance();
        realm.beginTransaction();
        if (refresh == true) {
            try {
                ArrayList<JSONObject> items = new ArrayList<JSONObject>();
                JSONArray DatabaseResults = new JSONArray();
                do {
                    String url = LeancloudAPIBaseURL + "/1.1/classes/ReBuildingRoom?order=BLE&limit=1000&skip=" + items.size();
                    JSONObject response = DatabaseHelper.LCSearch(url);
                    DatabaseResults = response.getJSONArray("results");
                    for (int i = 0; i < DatabaseResults.length(); i++) {
                        JSONObject checkLog = DatabaseResults.getJSONObject(i);
                        String buildingID = checkLog.getString("BuildingID");
                        String roomID = checkLog.getString("RoomID");
                        Building building = realm.where(Building.class).equalTo("ID", buildingID).findFirst();
                        Room room = realm.where(Room.class).equalTo("ID", roomID).findFirst();
                        building.Rooms.add(room);
                        room.Location = building;
                        items.add(checkLog);
                    }
                } while (DatabaseResults.length() != 0);
            } catch (Exception ignored) {
                realm.cancelTransaction();
                realm.close();
                return;
            }
        }
        realm.commitTransaction();
        realm.close();
    }
}

File: Qin-Backup/Qin/Qin-ios/Qin-ios/Util/Model/ReBuildingRoom.swift
//
//  ReBuildingRoom.swift
//  Qin-ios
//
//  Created by 杨璨 on 2020/2/29.
//  Copyright © 2020 canyang. All rights reserved.
//

import Foundation
import RealmSwift

public class ReBuildingRoom {
    public static func CreateBuildingRoom(refresh: Bool) {
        if refresh == true {
            var DatabaseResults = [[String:Any?]]()
            var getDBR = [[String:Any?]]()
            repeat {
                getDBR = [[String:Any?]]()
                let url = DatabaseHelper.LeancloudAPIBaseURL + "/1.1/classes/ReBuildingRoom?order=ID&limit=1000&skip=\(DatabaseResults.count)"
                if let response = DatabaseHelper.LCSearch(searchURL: url) {
                    getDBR = response["results"] as! [[String:Any?]]
                    DatabaseResults.append(contentsOf: getDBR)
                }
            } while getDBR.count != 0
            Building.GetAll(refresh: true)
            Room.GetAll(refresh: true)
            autoreleasepool {
                let realm = try! Realm()
                for checkLog in DatabaseResults {
                    realm.beginWrite()
                    let buildingID = checkLog["BuildingID"] as! String
                    let roomID = checkLog["RoomID"] as! String
                    if let building = realm.objects(Building.self).first(where: {$0.ID == buildingID}) {
                        if let room = realm.objects(Room.self).first(where: {$0.ID == roomID}) {
                            room.Location = building
                            building.Rooms.append(room)
                        }
                    }
                    try! realm.commitWrite()
                }
            }
        }
    }
}
